File: Juan/P4/Ruby/Napakalaki/lib/Napakalaki.rb
#encoding: utf-8

#Versión 3.0
require 'singleton'
require_relative 'Player.rb'
require_relative 'CardDealer.rb'
require_relative 'Dice.rb'
require_relative 'Monster.rb'
require_relative 'CultistPlayer.rb'
require_relative 'Cultist.rb'

module NapakalakiGame


class Napakalaki 
   include Singleton
   
    #@@instance = nil
#    Constructor

#    def initialize(currentPlayer,players,dealer,currentMonster,unusedMonsters)
#        @currentPlayer = currentPlayer
#        @players = players
#        @dealer = dealer
#        @currentMonster = currentMonster
#        @unusedMonsters = unusedMonsters
#    end


    def initialize()
        @currentPlayer 
        @players 
        @dealer = CardDealer.instance()
        @currentMonster  
    end

    attr_accessor :currentPlayer , :player, :dealer, :currentMonster
    
    ## getCurrentPlayer
    ## getCurrentMonster
    #attr_reader :currentPlayer
    #attr_reader :currentMonster
    
    #private_class_method :new
    
#    Inicializa el array de jugadores que contiene Napakalaki, creando tantos jugadores como
#    elementos haya en names, que es el array de String que contiene el nombre de losaa
#    jugadores.
   
    def initPlayers(names)
         @players = Array.new
           names.each do |n|
           @players <<  Player.new(n)
           
          end
         
    end
   
    private 
    def nextPlayer()
        if(@currentPlayer == nil)
            numero = rand(@players.size)
            puts @players.size
            puts @players.fetch(0)
            puts @players.fetch(1)
            @currentPlayer = @players[numero]
           
            aux=Array.new
            
            aux << @currentPlayer
            
            @players.each do |p|
                if ( p != @currentPlayer)
                    aux << p
                end
            end
            @players= aux
            
            return @currentPlayer
            
        else
            i=0
            cambiado= false
            while( i< @players.size)
                if(@currentPlayer == @players[i])
                    if(cambiado == false)
                        if(i ==(@players.size-1))
                            @currentPlayer = @players[0]
                            cambiado=true
                        else
                            @currentPlayer = @players[i+1]
                            cambiado=true
                        end
                    end
                end
                i = i+1
            end
            
        end
        return @currentPlayer
    end
    

    public
    def getCurrentPlayer
        
        return @currentPlayer
    end
    
    def getCurrentMonster
        return @currentMonster
    end
    
    private
    def nextTurnAllowed()
        sig = false
        puts 'Este es el currentPlayer'
        puts @currentPlayer
        
        if(@currentPlayer.validState)        
            sig=true
        return sig
        
        end
    end
    
    
    private
    def setEnemies()
        enemigo =Player.new(@players[0])
        #enemigo = @players[0]
        
        @players.each do |p|
            
            loop do 
                 numero = rand(@players.size)
                  
                 enemigo.setEnemy(@players[numero])
                 
            break if (enemigo != p)
            end
        end 
    end
    
#    private 
#    def self.getInstance()
#        @@instance = Napakalaki.instance
#    end
#   

    
    public
    def developCombat()
        combatResult = @currentPlayer.combat(@currentMonster)
        
        if combatResult == CombatResult::LOSEANDCONVERT
              dealer=CardDealer.instance 
              cultistPlayer = CultistPlayer.new(@currentPlayer,dealer.nextCultist())
              
            @players.each do |p|
                if (p == @currentPlayer)
                    p = cultistPlayer
                end
                
                if p.enemy == @currentPlayer
                    p.enemy = cultistPlayer
                end
            end
              
            @currentPlayer = cultistPlayer
        end
        
        return combatResult
    end
    

    
    def discardVisibleTreasures (treasures)
        treasures.each do |t|
            @currentPlayer.discardVisibleTreasures(t)
            @dealer.giveTreasureBack(t)
        end
    end
    
    def discardHiddenTreasures(treasures)
        treasures.each do |t|
            @currentPlayer.discardHiddenTreasures(t)
            @dealer.giveTreasureBack(t)
        end
    end
    
   

    def makeTreasureVisible(treasures)

        treasures.each do |t|
            @currentPlayer.discardHiddenTreasures(t)
            @dealer.giveTreasureBack(t)
        end
        
    end
    
    def initGame(players)
        initPlayers(players)
        setEnemies()
        @dealer.initCards
        nextPlayer
        nextTurn()
            

        
    end
    

    def nextTurn()
        stateOK = nextTurnAllowed
        
        if(stateOK == true)
            puts 'entra stateOK'
            @currentMonster = @dealer.nextMonster
            @currentPlayer = nextPlayer
            dead = @currentPlayer.isDead()
            
            if(dead == true)
                @currentPlayer.initTreasures
            end
        end
        
        return stateOK
    end
    
    def endOfGame (result)
        
       final = false
       if (result.eql?([CombatResult::WINGAME]))
            final = true
            
       end
            return final
   end
   
    
    
end

end

File: Juan/P5/java/Napakalaki/src/NapakalakiGame/Napakalaki.java
package NapakalakiGame;

import java.util.ArrayList;
import java.util.Random;

/**
 *
 * @author juane
 * @version 3.1 clase singleton
 */
public class Napakalaki {

    private static final Napakalaki instance = new Napakalaki();

    /* Relaciones con las clases */
    private Monster currentMonster ; // Relación con los monstruos    
    private CardDealer dealer = CardDealer.getInstance() ; // Relación con CardDealer
    private Player currentPlayer; // Relación con Player
    private ArrayList<Player> players = new ArrayList(); // Relación con Player

    /**
     * Constructor privado
     */
    private Napakalaki() {

    }

    /**
     * Método initPlayers(ArrayList<String> names
     *
     * @param names
     */
    private void initPlayers(ArrayList<String> names) {
        for (String n : names) {
            this.players.add(new Player(n));
        }
    }

    /**
     * Método nextPlayer Decide qué jugador es el siguiente en jugar.
     *
     * @return
     */
    private Player nextPlayer() {
       
       
        if (currentPlayer == null) {
            Random rand = new Random();
            int numero = rand.nextInt(players.size());

            currentPlayer = players.get(numero);
            ArrayList<Player> aux = new ArrayList();

            aux.add(currentPlayer);

            for (Player p : players) {
                if (p != currentPlayer) {
                    aux.add(p);
                }
            }
            players = aux;
            
            
            this.currentPlayer = aux.get(numero);
            return currentPlayer;
            
        } 
        else {
            boolean cambiado = false;
            for (int i = 0; i < players.size() && !cambiado; i++) {
               if(currentPlayer == players.get(i)){
                    if (i == players.size() - 1) ///////si falla mirar aquí
                    {
                        System.out.print("vuelve al primero" + i);
                        currentPlayer = players.get(0);
                        cambiado=true;
                    } 
                    else {
                        System.out.print("Entra al segundo" + i);
                        currentPlayer = players.get(i + 1);
                        cambiado=true;
                    }
                
                
                }
                
            }

            return currentPlayer;
        }
    }

    /**
     * Método nextTurnAllowed()
     *
     * @return
     */
   private boolean nextTurnIsAllowed() {

        boolean next = false;
        if(currentPlayer == null)
            next = true;
        else{
//            if (this.currentPlayer.validState()) {
//                next = true;
            next = this.currentPlayer.validState();
                return next;
            }
//            else{
//                return next;
//            }
        return next;
        }
      

    /**
     * Método setEnemies()
     */
    private void setEnemies() {

        Player enemigo = players.get(0);

        Random rand = new Random();
        for (Player p : players) {
            do {
                int numero = rand.nextInt(players.size());
                enemigo.setEnemy(players.get(numero));

            } while (enemigo.enemy == p);
        }

    }

    /**
     * Método getInstance()
     *
     * @return
     */
    public static Napakalaki getInstance() {
       
        return instance;
    }

    /**
     * Método developCombat()
     *
     * @return
     * @throws java.lang.CloneNotSupportedException
     */
    public CombatResult developCombat() {

        CombatResult combatResult = this.currentPlayer.combat(this.currentMonster);
        if (combatResult == CombatResult.LOSEANDCONVERT){
            
            CultistPlayer cultistPlayer = new CultistPlayer(currentPlayer,dealer.nextCultist());
            
            for(int i=0; i< players.size(); i++){
                if(players.get(i) == currentPlayer){
                   players.add(i, cultistPlayer);
                   players.remove(players.get(i+1));
                    
                    
                }
                if(players.get(i).enemy == currentPlayer)
                    players.get(i).enemy = cultistPlayer;
                    
            }
             currentPlayer = cultistPlayer;
        }
       
        return combatResult;
    }

    /**
     * Método discardVisibleTreasures(ArrayList<Treasure> treasures)
     *
     * @param treasures
     */
    public void discardVisibleTreasures(ArrayList<Treasure> treasures) {
        
        if (currentPlayer.getVisibleTreasures() != null && treasures != null){
            for (Treasure t : treasures) {
                this.currentPlayer.discardVisibleTreasure(t);

                this.dealer.giveTreasureBack(t);
            }
        }
    }

    /**
     * Método discardHiddenTreasures(ArrayList<Treasure> treasures)
     *
     * @param treasures
     */
    public void discardHiddenTreasures(ArrayList<Treasure> treasures) {
        if(currentPlayer.getVisibleTreasures() !=null && treasures != null){
            for (Treasure t : treasures) {
                this.currentPlayer.discardHiddenTreasure(t);

                this.dealer.giveTreasureBack(t);
            }
        }
    }

    /**
     * Método makeTreasuresVisible(ArrayList<Treasure> treasures)
     *
     * @param treasures
     */
    public void makeTreasuresVisible(ArrayList<Treasure> treasures) {

        for (Treasure t : treasures) {
            this.currentPlayer.makeTreasureVisible(t);
        }
    }

    /**
     * Método initGame(ArrayList<String> players)
     *
     * @param players
     */
    public void initGame(ArrayList<String> players) {
        
        this.initPlayers(players);
        this.setEnemies();
        dealer.initCards();
        this.nextTurn();
        

    }

    /**
     * Método getCurrentPlayer()
     *
     * @return
     */
    public Player getCurrentPlayer() {

        return this.currentPlayer; //cambiar
    }

    /**
     * Método getCurrentMonster()
     *
     * @return
     */
    public Monster getCurrentMonster() {
        return this.currentMonster;
    }

    /**
     * Método nextTurn()
     *
     * @return
     */
    public boolean nextTurn() {
        //MIRAR
        boolean stateOK;
        stateOK = this.nextTurnIsAllowed();

        System.out.print(" nextTurn stateOK\n");
        if (stateOK) {
            boolean dead;
            this.currentMonster = this.dealer.nextMonster();
            this.currentPlayer = this.nextPlayer();
            dead = this.currentPlayer.isDead();
            System.out.print(" Entra stateOK\n");
            if (dead) {
                System.out.print(" Entra Esta muerto\n");
                this.currentPlayer.initTreasures();
            }
        } //stateOK = this.currentPlayer.validState(); //PODRIA SER ASÍ
        else {
            this.currentMonster = this.dealer.nextMonster();
        }
        return stateOK;
    }

    /**
     * Método endOfGame(CombatResult result)
     *
     * @param result
     * @return
     */
    public boolean endOfGame(CombatResult result) {

        boolean end = false;
        if (result.equals(CombatResult.WINGAME)) {
            end = true;
        }

        return end;
    }
}
